### PR TITLE
ClinixGo App

### PR DESCRIPTION
**Pull request by Strangeal**
## Full-Stack Capstone
**Built a responsive app that allows users to book appointments with doctors.**

> In this PR, we completed the following tasks:

#### Backend
- [x] Created a new rails API app.
- [x] Set it up with PostgreSQL.
- [x] Setup linters and rspec.
- [x] Created models for the `user`, `doctor` and `appointments`.
- [x] Added the `model attributes` and `associations`.
- [x] Added `spec` files and validations to the models.
- [x] Created `controllers` and `controller actions` for the models.
- [x] Added `API routes`.
- [x] Implemented `authentication` using `jwt`.
- [x] Implemented `authorization` using `cancancan` gem.
- [x] Created the API Documentation using `rswag` and `swagger-ui`.
- [x] Tested all API endpoints with `Postman`.

#### Frontend

- [x] Created a new react app.
- [x] Set it up using `wepback` and `jsbundle`.
- [x] Set up linters and installed all necessary packages.
- [x] Set up the `routes` to the various pages.
- [x] Setup the redux store.
    - [x] Setup a reducer to handle user authentication (login and register).
    - [x] Setup a reducer to handle:
          - Fetching all doctors from the API endpoint.
          - Creating a doctor.
          - Displaying the details of a particular doctor.
     - [x] Setup a reducer to the handle:
           - The creation of appointments.
           - The listing of all appointments of the current user.
- [x] Provided the store to the components.
- [x] Created a `Login` component to allow a user to log in to the application.
- [x] Created a `Register` component to allow a new user to register an account.
- [x] Created a `Navbar` component to allow a user navigate to different parts of the app.
- [x] Created a `Homepage` component to show all the doctors.
- [x] Created a component to display all appointments of a user.
- [x] Created a component to allow a `logged in` user to book an appointment with a doctor.
- [x] Created a component to allow a `logged in` user to `add` a doctor if the user is an `admin`.
- [x] Created a component to allow a `logged in` user to `delete` a doctor if the user is an `admin`.
- [x] Added a functionality to allow a user to `log out` of the app.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+
+### Usage
+
+`./bin/dev`
+
+### Run Test
+
+To run app test. run the following commad:
+
+`rspec`
```
Comment (Meltrust) [inline README.md:129]: - [ ] If I run your tests according to your instructions, I get this:

image

If these tests are passing for you, kindly add instructions on your readme about how to properly run them and make them pass.  If this is a bug in the tests, kindly fix it to ensure proper functionality. 👍 

Also, there are 2 tests that are "pending", kindly either remove it or implement it.+

That way, your app's integrity will be secured.

File: app/javascript/styles/doctors.css [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+.doc-content .name {
+  font-size: 0.8rem;
+  font-weight: 600;
+}
+
+/* .doc-content .contact {
+  font-size: 0.7rem;
+} */
```
Comment (Meltrust) [inline app/javascript/styles/doctors.css:14]: - [ ] Kindly, remove the commented-out code present in this file in order to ensure the repo's cleanliness. 👍

File: config/routes.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+Rails.application.routes.draw do
+  mount Rswag::Ui::Engine => '/api-docs'
+  mount Rswag::Api::Engine => '/api-docs'
+  root 'root#index'
+
+  namespace :api do
+    namespace :v1 do
+      resources :doctors, only: [:index, :show, :create, :update]
+      resources :users
+      resources :appointments
```
Comment (Meltrust) [inline config/routes.rb:10]: - [ ] OPTIONAL: Kindly, ensure you are not generating more routes than those that are needed.


Kindly check if `resources :users` is needed.  You are not using any of index, show, edit, create, update and destroy methods in your users controller.

Kindly check if 'resources :appointments' is needed without an `only:` argument.  


That way, your routes table will be simpler, and your app will be faster. 👍

**Review by Meltrust — CHANGES_REQUESTED** [review]
Hi @Strangeal,

Good job so far!
There are some issues that you still need to work on to prepare your project for the final evaluation, but you are almost there!






### To highlight:


-  App is working well✔️ 
-  Good readme ✔️ 
-  Beautiful frontend ✔️ 
-  Nice code organization and implementation ✔️ 


#### You are really close to finishing the Microverse program!! Keep it up! 👍👍👍

You can do it

### After implementing the requested changes, please submit another review request. ♻️
_Check the comments under the review._

_I strongly recommend you take them into account as they can improve your final evaluation._





Cheers and Happy coding!👏👏👏

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the previous reviews unless it is requested otherwise._**

------

File: app/javascript/components/Appointment.js [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+                type="submit"
+                className="mt-4 btn btn-success-cus d-block w-100 rounded-pill"
+              >
+                Book Appointment
+              </button>
+            </form>
+          </div>
+        </div>
+      </div>
+    </div>
```
Comment (mcrd25) [inline app/javascript/components/Appointment.js:152]: [OPTIONAL] When you see similar parts being repeated (like same style of div) you can consider simplifying code with loop or similar.

**Review by mcrd25 — APPROVED** [review]
Hi Team, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
- You've addressed concerns of previous reviewer ✔️ 

## Optional suggestions
- Tests don't seem uniform in sense that there are cases where you do positive and negative scenario tests in user_spec (which is good), but, you do not do the same for others. Also, you do tests in the other specs you don't do in user_spec. Consider this before moving on.
- Also, while I don't see any obvious blunders, always try to see how you can improve your code in terms of design patterns and best practices.

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by Strangeal** [comment]
Thank you @mcrd25

**Pull request merged**